File: Arrays/238. Product of Array Except Self.py
from typing import List

#brute force approach - T: O(n^2)
# class Solution:
#     def productExceptSelf(self, nums: List[int]) -> List[int]:
#         res = []
#         for i in range(len(nums)):
#             temp = 1
#             for j in range(len(nums)):
#                 if(j == i):
#                     continue
#                 temp *= nums[j]
#             res.append(temp)
#         return res
    
#optimal approach - T: O(n)
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        pass
        
if __name__ == "__main__":
    sol = Solution()
    nums = [1,2,3,4]
    print(sol.productExceptSelf(nums))

File: Arrays/238. Product of Array Except Self.cpp
#include <bits/stdc++.h> 
using namespace std;

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> res(nums.size());

        // brute force
        // for (int i = 0; i < nums.size(); i++) {
        //     int product = 1;
        //     for (int j = 0; j < nums.size(); j++) {
        //         if (j == i) continue;

        //         product = product * nums[j];
        //     }
        //     res[i] = product;
        // }


        // optimal approach 
        // calculate prefix
        int prefix = 1;
        for (int i = 0; i < nums.size(); i++) {
            res[i] = prefix;
            prefix *= nums[i];
        }

        // calculate the postfix
        int postfix = 1;
        for (int i = nums.size() - 1; i >= 0; i--) {
            res[i] *= postfix;  /* multiply postfix with prefix value*/
            postfix *= nums[i];
        }

        return res; 
    }
};


int main() {
    vector<int> nums = {1,2,3,4};
    Solution sol;
    vector<int> res = sol.productExceptSelf(nums);
    for (auto num: res) {
        cout << num;
    }
    return 0;
}